Custom sort function for InitContainersStatuses

lukaszo opened the pull request:
Order in init containers matters. Statues shoudln't be sorted by name.

lukaszo commented:
@smarterclayton can you please take a look at this?

lukaszo commented:
The order is wrong because of this sort: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/manager.go#L546

smarterclayton commented:
Hrm, that sort is probably wrong then.

smarterclayton commented:
We should probably fix the sort of statuses - it should always be in order of the init container initial positions.

lukaszo commented:
Ok, I will take a look  at it

smarterclayton commented:
Bumping to p1 since this results in inconsistent client experience with init containers.

lukaszo committed:
Custom sort function for InitContainersStatuses

Order in init container statuses should be the same as defined in pod.
Statues shoudln't be sorted by name.

lukaszo commented:
Ok, there was one more sort call. I changed it so InitContainerStatuses is not sorted now. With this changes kubectl shows correct status now.

Maybe in long term InitContainerStatuses should have different type which strictly forbids sorting. It would require much more changes though.

smarterclayton commented:
Well, the statuses still need to be sorted - they just need to be in a specific sort that is not the same as app containers.  We depend on the sort order of init container statuses in some cases.

smarterclayton commented:
Basically the sort we need to use for init containers is:

"order that init containers appear in the pod"

lukaszo commented:
Ah, I thought that it will be in correct order now. My bad. 
Ok, so I see two solutions here:
* write custom sort function which will use data from pod.Spec.InitContainers to get correct order
* add `index` field to ContainerStatus struct and create Interface to sort by this index

This `index` field would not make any sense for pod containers so I think first option is better. What do you think?  Or do you see better approach?

smarterclayton commented:
I think the first approach makes sense - you should be able to swap the two positions as you walk the status array (find the first name, swap it into the first position, find the second name, swap it into the second position) in classic bubble sort.

lukaszo commented:
It's getting more complicated. In normalizeStatus from https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/manager.go#L507  I have only access to api.PodStatus and it's not enough to sort InitContainers.
Any idea how to solve it? Passing pod.Spec  through all functions sound like a bad idea :/

smarterclayton commented:
How many places call normalizeStatus?  The only use is see has the pod spec
available.

On Jun 6, 2016, at 6:34 PM, Łukasz Oleś <notifications@github.com> wrote:

It's getting more complicated. In normalizeStatus from
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/manager.go#L507
I have only access to api.PodStatus and it's not enough to sort
InitContainers.
Any idea how to solve it? Passing pod.Spec through all functions sound like
a bad idea :/

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/26829#issuecomment-224109634>,
or mute the thread
<https://github.com/notifications/unsubscribe/ABG_py0f7t5JM9r_dKf3Nz3UzVBeIbykks5qJKCNgaJpZM4IuAeQ>
.

lukaszo commented:
It's called from `updateStatusInternal` and from `needsReconcile` in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/manager.go `needReconcile` is called from syncBatch

lukaszo commented:
If passing pod to normalizeStatus is ok for you then there is no problem.

smarterclayton commented:
It looks like both cases you can pass pod spec in, that is ok with me.

lukaszo commented:
I added unit test for sort function. Seems to work. Are there any other cases to check?

smarterclayton commented on pkg/kubelet/types/types.go:
```diff
@@ -68,6 +68,21 @@ func (s SortedContainerStatuses) Less(i, j int) bool {
 	return s[i].Name < s[j].Name
 }
 
+// Sorts InitContainerStatuses
```
Minor thing, but we try to stay consistent to Godoc style which is `SortInitContainerStatuses ensures that statuses are in the order that their init container appears in the pod spec."  I.e. one sentence minimum, starts with the name of the function.

smarterclayton commented:
One minor comment, if you can fix that and squash I'll apply the label.  Thanks!

lukaszo commented:
Ok,  copy/paste of your text

smarterclayton commented:
Thanks, squash

lukaszo commented:
done

smarterclayton commented on pkg/kubelet/types/types.go:
```diff
@@ -68,6 +68,22 @@ func (s SortedContainerStatuses) Less(i, j int) bool {
+	containers := p.Spec.InitContainers
+	current := 0
+	for _, container := range containers {
+		for j := current; j < len(statuses); j++ {
```
Hrm, if someone passes a statuses where container.Name is not in statuses, will this loop forever?

lukaszo replied:
Hmm, I don't see how. Could you provide test case for unit test?

smarterclayton replied:
Never mind, it looks fine.  Just wanted to be extra sure.

smarterclayton commented:
Looks like the unit test fails to build

lukaszo commented:
Yes, I fixed unit tests. Now I'm trying to find the issue with GCE Node e2e tests

smarterclayton commented:
Node tests were not you - was a bug today:

@k8s-bot test this issue #26431

On Tue, Jun 7, 2016 at 3:51 PM, Łukasz Oleś <notifications@github.com>
wrote:

> Yes, I fixed unit tests. Now I'm trying to find the issue with GCE Node
> e2e tests
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26829#issuecomment-224393879>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p2zvUvApEaX_OHdB0FrIRuDDYVyOks5qJcu7gaJpZM4IuAeQ>
> .
>

lukaszo commented:
@smarterclayton can you add label LGTM? Or do you have any more comments?

smarterclayton commented:
Was waiting for the submit queue to get back to normal.  Marked now, thanks again.

lukaszo commented:
@smarterclayton thank you for your help and patience  :)
